cirq/ops/common_gates.py

```diff
@@ -490,17 +490,48 @@ def measure_each(*qubits: raw_types.QubitId,
 T = Z**0.25
 
 
-class HGate(gate_features.CompositeGate,
+class HGate(eigen_gate.EigenGate,
             gate_features.TextDiagrammable,
-            gate_features.ReversibleEffect,
-            gate_features.KnownMatrix,
             gate_features.SingleQubitGate,
             gate_features.QasmConvertibleGate):
     """180 degree rotation around the X+Z axis of the Bloch sphere."""
 
-    def text_diagram_info(self, args: gate_features.TextDiagramInfoArgs
-                          ) -> gate_features.TextDiagramInfo:
-        return gate_features.TextDiagramInfo(('H',))
+    def __init__(self, *,  # Forces keyword args.
+                 half_turns: Optional[Union[value.Symbol, float]] = None,
+                 rads: Optional[float] = None,
+                 degs: Optional[float] = None) -> None:
+        """Initializes the gate.
+
+        At most one angle argument may be specified. If more are specified,
+        the result is considered ambiguous and an error is thrown. If no angle
+        argument is given, the default value of one half turn is used.
+
+        Args:
+            half_turns: The relative phasing of H's eigenstates, in half_turns.
+            rads: The relative phasing of H's eigenstates, in radians.
+            degs: The relative phasing of H's eigenstates, in degrees.
+        """
+        super().__init__(exponent=value.chosen_angle_to_half_turns(
+            half_turns=half_turns,
+            rads=rads,
+            degs=degs))
+
+    def _canonical_exponent_period(self) -> Optional[float]:
+        return 2
+
+    def _with_exponent(self,
+                       exponent: Union[value.Symbol, float]) -> 'RotXGate':
+        return HGate(half_turns=exponent)
+
+    def _eigen_components(self):
+        return [
+            (0, np.array([[0.5, 0.5], [0.5, 0.5]])),
```

Comment:
These are the eigencomponents for X. For H, the first one should be

[[3 + 2 sqrt(2), 1 + sqrt(2)], [1 + sqrt(2), 1]] / (2 * (2 + sqrt(2))

and something similar for the second.

Reply:
This is what I have 

```
def _eigen_components(self):
        two_plus_sqrt_two = 2 * (2 + np.sqrt(2))

        component = np.array([[(3 + 2 * np.sqrt(2)) / two_plus_sqrt_two, (1 + np.sqrt(2)) /
                               two_plus_sqrt_two], [(1 + np.sqrt(2)) / two_plus_sqrt_two, (1) / (2 * two_plus_sqrt_two)]])

        return [(0, component), (1, component), ]
```